modules/lib/core/index.d.ts

```diff
@@ -52,23 +67,115 @@ export interface PublishInfo {
     first?: string;
 }
 
-export interface Component<Config extends object = object, Regions extends Record<string, Region> = Record<string, Region>> {
-    config: Config;
-    descriptor: string;
+export interface FragmentComponent {
+    type: 'fragment'
+    fragment: string;
     path: string;
-    type: 'page' | 'layout' | 'part';
+}
+
+export interface LayoutComponent<
+    Descriptor extends ComponentDescriptor = ComponentDescriptor,
+    Config extends NestedRecord = NestedRecord,
+    Regions extends
+        Record<string, Region<(FragmentComponent | PartComponent | TextComponent)[]>> = 
+        Record<string, Region<(FragmentComponent | Part          | TextComponent)[]>>
+> {
+    type: 'layout'
+    descriptor: Descriptor
+    config: Config
+    path?: string // Missing in fragmentPreview https://github.com/enonic/xp/issues/10116
+    regions: Regions;
+}
+type LayoutDescriptor = keyof XpLayoutMap;
+type Layout = LayoutDescriptor extends any // this lets us iterate over every member of the union
+    ? LayoutComponent<LayoutDescriptor>
+    : never;
+
+export interface PartComponent<
+    Descriptor extends ComponentDescriptor = ComponentDescriptor,
+    Config extends NestedRecord = NestedRecord
+> {
+    type: 'part'
+    descriptor: Descriptor
+    config: Config
+    path?: string // Missing in fragmentPreview https://github.com/enonic/xp/issues/10116
+}
+type PartDescriptor = keyof XpPartMap;
+type Part = PartDescriptor extends any // this lets us iterate over every member of the union
+    ? PartComponent<PartDescriptor>
+    : never;
+
+export interface PageComponent<
+    Descriptor extends ComponentDescriptor = ComponentDescriptor,
+    Config extends NestedRecord = NestedRecord,
+    Regions extends
+        Record<string, Region<(FragmentComponent | LayoutComponent | PartComponent | TextComponent)[]>> = 
+        Record<string, Region<(FragmentComponent | Layout          | Part          | TextComponent)[]>>
+> {
+    type: 'page'
+    descriptor: Descriptor
+    config: Config
+    path: '/'
     regions: Regions;
 }
+type PageDescriptor = keyof XpPageMap;
+type Page = PageDescriptor extends any // this lets us iterate over every member of the union
+    ? PageComponent<PageDescriptor>
+    : never;
+
+export interface TextComponent {
+    type: 'text'
+    path: string
+    text: string
+}
+
+export type Component<
+    Config extends NestedRecord = NestedRecord,
+    Regions extends Record<string, Region> = Record<string, Region>,
+    Descriptor extends ComponentDescriptor = LayoutDescriptor | PageDescriptor | PartDescriptor,
```

Comment:
@tajakobsen Had to add a third param here, to match the generics for LayoutComponent, PageComponent and PartComponent

Reply:
I would reorder the type parameters. Since `Regions` are not used for every type in the union it should come last. 
 1. Config
 2. Descriptor
 3. Regions

Reply:
This is also consistent with the order of type parameters in `Content`

Reply:
I would definetly have Regions as the third param, it's a breaking change though.

(I would make descriptor the first key though). Just a config would not tell whether the component is Layout, Page or Part...